File: monads-java/src/main/java/com/horbowicz/monads/java/io/IO.java
package com.horbowicz.monads.java.io;

import java.util.function.Function;

public final class IO<T> {

  private final Operation<T> lazyOperation;

  private IO(Operation<T> lazyOperation) {
    this.lazyOperation = lazyOperation;
  }

  public <V> IO<V> map(Function<T, V> function) {
    return IO.of(() -> function.apply(run()));
  }

  public <V> IO<V> flatMap(Function<T, IO<V>> function) {
    return IO.of(() -> function.apply(run()).run());
  }

  public T run() {
    return lazyOperation.execute();
  }

  public interface Operation<V> {
    V execute();
  }

  public static <T> IO<T> of(Operation<T> lazyOperation) {
    return new IO<>(lazyOperation);
  }
}

File: monads-scala/src/main/scala/com/horbowicz/monads/scala/io/IO.scala
package com.horbowicz.monads.scala.io

final class IO[A](private val a: () => A) {
  def flatMap[B](f: A => IO[B]) = new IO(() => f(this.run()).run())

  def map[B](f: A => B): IO[B] = new IO(() => f(this.run()))

  def run(): A = a()
}

object IO {
  def apply[A](a: => A) = new IO(() => a)
}
